src/Common/FileCache.cpp

```diff
@@ -0,0 +1,609 @@
+        queue_size += 1;
+
+    auto is_overflow = [&]
+    {
+        return (current_size + size - removed_size > max_size)
+            || (max_element_size != 0 && queue_size > max_element_size);
+    };
+
+    std::vector<FileKeyAndOffset> to_evict;
+
+    auto key_it = queue.begin();
+    while (is_overflow() && key_it != queue.end())
+    {
+        const auto [key, offset] = *key_it++;
+
+        auto * cell = getCell(key, offset, cache_lock);
+        if (!cell)
+            throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+                "Cache became inconsistent. Key: {}, offset: {}", keyToStr(key), offset);
+
+        size_t cell_size = cell->size();
+
+        /// It is guaranteed that cell is not removed from cache as long as
+        /// pointer to corresponding file segment is hold by any other thread.
+
+        if (cell->releasable())
+        {
+            switch (cell->file_segment->download_state)
+            {
+                case FileSegment::State::DOWNLOADED:
+                {
+                    /// Cell will actually be removed only if
+                    /// we managed to reserve enough space.
+
+                    to_evict.emplace_back(key, offset);
+                    break;
+                }
+                case FileSegment::State::DOWNLOADING:
+                {
+                    remove(key, offset, cache_lock);
+                    break;
+                }
+                default:
+                {
+                    remove(key, offset, cache_lock);
+                    break;
+                }
+            }
+
+            removed_size += cell_size;
+            --queue_size;
+        }
+    }
+
+    if (is_overflow())
+        return false;
+
+    if (cell_for_reserve && !cell_for_reserve->queue_iterator)
+        cell_for_reserve->queue_iterator = queue.insert(queue.end(), std::make_pair(key_, offset_));
+
+    for (auto & [key, offset] : to_evict)
+        remove(key, offset, cache_lock);
+
+    current_size += size - removed_size;
+    if (current_size > (1ull << 63))
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Cache became inconsistent. There must be a bug");
+
+    return true;
+}
+
+void LRUFileCache::remove(
+    Key key, size_t offset, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+{
+    LOG_TEST(log, "Remove. Key: {}, offset: {}", keyToStr(key), offset);
+
+    auto * cell = getCell(key, offset, cache_lock);
+    if (!cell)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "No cache cell for key: {}, offset: {}", keyToStr(key), offset);
+
+    if (cell->queue_iterator)
+        queue.erase(*cell->queue_iterator);
+
+    auto & offsets = files[key];
+    offsets.erase(offset);
+
+    auto cache_file_path = path(key, offset);
+    if (fs::exists(cache_file_path))
+    {
+        try
+        {
+            fs::remove(cache_file_path);
+
+            if (startup_restore_finished && offsets.empty())
```

Comment:
Add assert?

Reply:
No, in case `startup_restore_finished==false` we might have: `restore() -> tryReserve() -> remove()`. Because when information is loaded base on files at startup we also check whether there is still enough space in cache as it's size could be changed by user in config..
And we check `if (startup_restore_finished && offsets.empty())` because not all is loaded yet.